File: frontend/src/types/task.ts
export type TaskStatus = "todo" | "in_progress" | "done";
export type TaskPriority = "low" | "medium" | "high";
export type TaskSortField = "due_date" | "priority" | "created_at" | "title";
export type SortOrder = "asc" | "desc";

/** FastAPI の TaskRead に対応する。 */
export type Task = {
  id: number;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  /** ISO8601（オフセット付き） */
  due_date: string | null;
  estimated_minutes: number | null;
  completed_at: string | null;
  /** 分解元のタスク。分解で作られた小タスクのみ値が入る */
  parent_task_id: number | null;
  created_at: string;
  updated_at: string;
};

export type TaskInput = {
  title: string;
  description?: string | null;
  status?: TaskStatus;
  priority?: TaskPriority;
  /** "2026-09-30T23:59" のようなオフセット無しの値を送ると Backend が Asia/Tokyo として解釈する */
  due_date?: string | null;
  estimated_minutes?: number | null;
};

export type TaskFilters = {
  statuses: TaskStatus[];
  priorities: TaskPriority[];
  keyword: string;
  sortBy: TaskSortField;
  order: SortOrder;
};

export const TASK_STATUS_LABEL: Record<TaskStatus, string> = {
  todo: "未着手",
  in_progress: "進行中",
  done: "完了",
};

export const TASK_PRIORITY_LABEL: Record<TaskPriority, string> = {
  high: "高",
  medium: "中",
  low: "低",
};

export const TASK_SORT_LABEL: Record<TaskSortField, string> = {
  due_date: "期限",
  priority: "優先度",
  created_at: "作成日",
  title: "タイトル",
};

File: backend/app/schemas/task.py
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import TaskPriority, TaskStatus
from app.schemas.common import MAX_DESCRIPTION, AwareDatetime


class TaskCreate(BaseModel):
    """タスク作成の入力。通常UI と LLM Tool の両方がこのスキーマで検証される。"""

    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=MAX_DESCRIPTION)
    status: TaskStatus = TaskStatus.TODO
    priority: TaskPriority = TaskPriority.MEDIUM
    due_date: AwareDatetime | None = None
    estimated_minutes: int | None = Field(default=None, gt=0, le=60 * 24)
    parent_task_id: int | None = Field(
        default=None, gt=0, description="分解元のタスクID"
    )


class TaskUpdate(BaseModel):
    """部分更新。未指定の項目は変更しない（null を明示すればクリアできる）。"""

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=MAX_DESCRIPTION)
    status: TaskStatus | None = None
    priority: TaskPriority | None = None
    due_date: AwareDatetime | None = None
    estimated_minutes: int | None = Field(default=None, gt=0, le=60 * 24)


class TaskRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str | None
    status: TaskStatus
    priority: TaskPriority
    due_date: datetime | None
    estimated_minutes: int | None
    completed_at: datetime | None
    parent_task_id: int | None
    created_at: datetime
    updated_at: datetime


class TaskSortField(StrEnum):
    DUE_DATE = "due_date"
    PRIORITY = "priority"
    CREATED_AT = "created_at"
    TITLE = "title"


class SortOrder(StrEnum):
    ASC = "asc"
    DESC = "desc"


class TaskSearchParams(BaseModel):
    """一覧・検索の条件。LLM の search_tasks Tool でもそのまま使う。"""

    statuses: list[TaskStatus] | None = None
    priorities: list[TaskPriority] | None = None
    keyword: str | None = Field(default=None, max_length=200)
    due_from: AwareDatetime | None = None
    due_to: AwareDatetime | None = None
    sort_by: TaskSortField = TaskSortField.DUE_DATE
    order: SortOrder = SortOrder.ASC
    limit: int = Field(default=100, ge=1, le=500)
    offset: int = Field(default=0, ge=0)
